File: 028_Dota2Senate.py
"""
Topic: Queue / Simulation
Dota2 Senate: simulate bans with two queues of indices
Time Complexity: O(n)
Space Complexity: O(n)
"""

from collections import deque

def predict_party_victory(senate: str) -> str:
    n = len(senate)
    rq = deque()
    dq = deque()
    for i, c in enumerate(senate):
        if c == 'R': rq.append(i)
        else: dq.append(i)
    while rq and dq:
        r = rq.popleft(); d = dq.popleft()
        if r < d:
            rq.append(r + n)
        else:
            dq.append(d + n)
    return 'Radiant' if rq else 'Dire'


if __name__ == '__main__':
    print(predict_party_victory('RD'))   # Radiant
    print(predict_party_victory('RDD'))  # Dire

File: 028_Dota2Senate.cpp
// Topic: Queue / Simulation
// Dota2 Senate: Predict which party (Radiant or Dire) will win using queue simulation.
// Time Complexity: O(n)
// Space Complexity: O(n)

#include <bits/stdc++.h>
using namespace std;

string predictPartyVictory(string senate)
{
    int n = senate.size();
    queue<int> r, d;
    for (int i = 0; i < n; ++i)
    {
        if (senate[i] == 'R')
            r.push(i);
        else
            d.push(i);
    }
    while (!r.empty() && !d.empty())
    {
        int ri = r.front();
        r.pop();
        int di = d.front();
        d.pop();
        if (ri < di)
            r.push(ri + n);
        else
            d.push(di + n);
    }
    return r.empty() ? string("Dire") : string("Radiant");
}

int main()
{
    cout << predictPartyVictory("RD") << '\n';   // Radiant
    cout << predictPartyVictory("RDD") << '\n';  // Dire
    cout << predictPartyVictory("RDRDR") << '\n';
    return 0;
}
